src/Compilers/Core/Portable/VersionHelper.cs

```diff
@@ -72,37 +72,36 @@ private static bool TryParse(string s, bool allowWildcard, ushort maxValue, out
                 }
             }
 
-            if (hasWildcard)
+            version = new Version(values[0], values[1], values[2], values[3]);
+            return true;
+        }
+
+        /// <summary>
+        /// If build and/or revision numbers are 65535 they are replaced with time-based values.
+        /// </summary>
+        public static Version GenerateVersionFromPatternAndCurrentTime(Version pattern)
+        {
+            if (pattern == null || pattern.Revision != ushort.MaxValue)
             {
-                int seconds = ((int)(DateTime.Now.TimeOfDay.TotalSeconds)) / 2;
-                if (seconds > (int)maxValue)
-                {
-                    version = null;
-                    return false;
-                }
-                values[3] = (ushort)seconds;
+                return pattern;
+            }
 
-                if (elements.Length == 3)
-                {
-                    TimeSpan days = DateTime.Today - new DateTime(2000, 1, 1);
-                    int build = (int)days.TotalDays;
+            int revision = (int)DateTime.Now.TimeOfDay.TotalSeconds / 2;
 
-                    if (build < 0 || build > (int)maxValue)
-                    {
-                        //alink would generate an error here saying "Cannot auto-generate build and 
-                        //revision version numbers for dates previous to January 1, 2000." Without
-                        //some refactoring here to relay the date problem, Roslyn
-                        //will generate an inaccurate error about the version string being of the wrong format.
+            // 24 * 60 * 60 / 2 = 43200 < 65534
+            Debug.Assert(revision < 0xffff);
 
-                        version = null;
-                        return false;
-                    }
-                    values[2] = (ushort)build;
-                }
-            }
+            if (pattern.Build == ushort.MaxValue)
+            {
+                TimeSpan days = DateTime.Today - new DateTime(2000, 1, 1);
+                int build = Math.Min(ushort.MaxValue, (int)days.TotalDays);
```

Comment:
If you set the current date to +150 years it will happen. Nobody really does that but it's not unreachable. I think caping it to 65535 is fine.